File: Leetcode/python-solutions/is-anagram.py
class Solution(object):
    def isAnagram(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        # first solution
        # works but the runtime and memory usages are midway
        # return sorted(s) == sorted(t)

        # second solution

        if len(s) != len(t):
            return False
        char_map = [0] * 26

        for i in range(len(s)):
            char_map[ord(s[i]) - ord('a')] += 1
            char_map[ord(t[i]) - ord('a')] -= 1

        return all(count == 0 for count in char_map)

        # a lot faster on runtime, can't seem to find a way
        # to reduce memory usage

File: Leetcode/typescript/is-anagram.ts
function isAnagram(s: string, t: string): boolean {
  if (s.length !== t.length) return false;

  const charMap: { [key: string]: number } = {};

  for (let i = 0; i < s.length; i++) {
    charMap[s[i]] = (charMap[s[i]] || 0) + 1;
    charMap[t[i]] = (charMap[t[i]] || 0) - 1;
  }

  for (const c in charMap) {
    if (charMap[c] !== 0) return false;
  }
  return true;
}
